### PR TITLE
With unit tests

### PR DESCRIPTION
**Pull request by npavlov**

### CODE REVIEW

File: cmd/gophermart/main.go [hunk 1]
```diff
@@ -1,3 +1,117 @@
...
+)
+
+const orderTopic = "orders"
+
+var (
+	ErrDatabaseNotConnected = errors.New("database is not connected")
+	ErrJWTisNotPorvided     = errors.New("JWT token is not provided")
+)
+
+//nolint:funlen
```
Comment (arttet) [inline cmd/gophermart/main.go:35]: Please remove it.

File: cmd/gophermart/main.go [hunk 2]
```diff
@@ -1,3 +1,117 @@
...
+	if cfg.JwtSecret == "" {
+		panic(ErrJWTisNotPorvided)
+	}
+
+	ctx, cancel := utils.WithSignalCancel(context.Background(), log)
+
+	dbManager := dbmanager.NewDBManager(cfg.Database, log).Connect(ctx).ApplyMigrations()
+	defer dbManager.Close()
+	if dbManager.DB == nil {
+		panic(ErrDatabaseNotConnected)
```
Comment (arttet) [inline cmd/gophermart/main.go:59]: Please don't panic

File: internal/handlers/auth/auth_test.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	"golang.org/x/crypto/bcrypt"
+
+	"github.com/npavlov/go-loyalty-service/internal/config"
+	handlers "github.com/npavlov/go-loyalty-service/internal/handlers/auth"
+	"github.com/npavlov/go-loyalty-service/internal/models"
+	testutils "github.com/npavlov/go-loyalty-service/internal/test_utils"
+)
+
+const (
+	password = "password123"
```
Comment (arttet) [inline internal/handlers/auth/auth_test.go:24]: https://github.com/onsi/ginkgo

File: internal/handlers/orders/orders_test.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+		// Add the order to the processing queue
+		err = mockOrders.AddOrder(context.Background(), orderID, userID)
+		require.NoError(t, err)
+
+		// Start processing in a separate goroutine
+		go mockOrders.ProcessOrders(context.Background())
+
+		// Wait for a moment to allow processing
+		time.Sleep(500 * time.Millisecond)
```
Comment (arttet) [inline internal/handlers/orders/orders_test.go:163]: This is a bad workaround.

File: internal/redis/model.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package redis
+
+import (
+	"context"
+	"time"
+)
+
+type MemStorage interface {
+	Ping(ctx context.Context) error
```
Comment (arttet) [inline internal/redis/model.go:9]: It should not be a part of `MemStorage`.

File: migrations/20241129140834_first_migration.sql [hunk 6]
```diff
@@ -0,0 +1,48 @@
+-- +goose Up
+-- create enum type "order_status"
+CREATE TYPE "order_status" AS ENUM ('NEW', 'PROCESSING', 'INVALID', 'PROCESSED');
```
Comment (arttet) [inline migrations/20241129140834_first_migration.sql:3]:
```suggestion
CREATE TYPE "order_status_type" AS ENUM ('NEW', 'PROCESSING', 'INVALID', 'PROCESSED');
```

File: migrations/20241129140834_first_migration.sql [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+-- create "users" table
+CREATE TABLE "users" (
+  "id" uuid NOT NULL DEFAULT gen_random_uuid(),
```
Comment (arttet) [inline migrations/20241129140834_first_migration.sql:6]: You need to enable the extension.

File: sql/schema.tmpl.sql [hunk 8]
```diff
@@ -0,0 +1,66 @@
+-- Enable the pgcrypto extension to use gen_random_uuid()
+CREATE EXTENSION IF NOT EXISTS pgcrypto;
```
Comment (arttet) [inline sql/schema.tmpl.sql:2]:
```suggestion
CREATE EXTENSION pgcrypto;
```

File: sql/schema.tmpl.sql [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                        user_id UUID REFERENCES users(id) ON DELETE CASCADE,
+                        order_num VARCHAR(255) UNIQUE NOT NULL,
+                        status order_status NOT NULL DEFAULT 'NEW', -- Default status is 'NEW'
+                        amount NUMERIC(10, 2) DEFAULT NULL,
+                        created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+                        updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP
+);
+
+-- Add trigger function to update the updated_at column
+CREATE OR REPLACE FUNCTION update_updated_at_column()
```
Comment (arttet) [inline sql/schema.tmpl.sql:28]: Are you sure about it?

**Review by arttet — APPROVED** [review]

File: cmd/gophermart/main.go (hunk 2 above)
Comment (npavlov) [reply cmd/gophermart/main.go:59]: fixed

File: internal/redis/model.go (hunk 5 above)
Comment (npavlov) [reply internal/redis/model.go:9]: fixed

File: migrations/20241129140834_first_migration.sql (hunk 7 above)
Comment (npavlov) [reply migrations/20241129140834_first_migration.sql:6]: it exists in schema but by default it's already enabled 


CREATE EXTENSION IF NOT EXISTS pgcrypto; in schema file

File: sql/schema.tmpl.sql (hunk 8 above)
Comment (npavlov) [reply sql/schema.tmpl.sql:2]: I can agree, receiving error:

executing statement: "CREATE EXTENSION pgcrypto;": pq: extension "pgcrypto" already exists

File: sql/schema.tmpl.sql (hunk 9 above)
Comment (npavlov) [reply sql/schema.tmpl.sql:28]: I agree, fixed

File: migrations/20241129140834_first_migration.sql (hunk 6 above)
Comment (npavlov) [reply migrations/20241129140834_first_migration.sql:3]: fixed

File: cmd/gophermart/main.go (hunk 1 above)
Comment (npavlov) [reply cmd/gophermart/main.go:35]: fixed

**Pull request closed without merging**